File: projects/08/NameTester.java
/* NameTester.java is a 'driver' to test class Name.
 * Begun by: Prof. Adams, for CS 214 at Calvin College.
 * Student: Justin Voss
 * Date: 3/10/22
 ************************************************************/
import java.io.*;
import java.util.Scanner;

public class NameTester  
{
  public static void main(String[] args)
  {
      Name aName = new Name("John", "Paul", "Jones");
//
      assert aName.getFirst().equals("John") ;
      assert aName.getMiddle().equals("Paul") ;
      assert aName.getLast().equals("Jones") ;
//
      System.out.println(aName);
      assert aName.toString().equals("John Paul Jones");


      // testing the mutators
      aName.setFirst("Chris");
      assert aName.getFirst().equals("Chris") ;

      aName.setMiddle("Robert");
      assert aName.getMiddle().equals("Robert") ;

      aName.setLast("Evans");
      assert aName.getLast().equals("Evans") ;
      System.out.println(aName);

      // testing the LFMI
      assert aName.lfmi().equals("Evans, Chris R");

      System.out.println("All tests passed!");

      // taking user input
      System.out.println("Enter a person's first, middle and last name:");
      aName.read();
      System.out.println(aName);

  }

}
  class Name
  {
     private String myFirst,
                    myMiddle,
                    myLast;

    public Name(String first, String middle, String last)
    {
        myFirst = first;
        myMiddle = middle;
        myLast = last;
    }

    public String getFirst()
   {
      return myFirst;
   }

   public String getMiddle()
   {
      return myMiddle;
   }

   public String getLast()
   {
      return myLast;
   }

   public String toString()
   {
      return myFirst + ' ' + myMiddle + ' ' + myLast;
   }
   // sets the first name
   void setFirst(String first)
   {
      myFirst = first;
   }
   // sets the middle name
   void setMiddle(String middle)
   {
      myMiddle = middle;
   }
   // sets the last name
   void setLast(String last)
   {
      myLast = last;
   }

   // this function puts the name in last, first, middle initial order
   String lfmi() 
   {
      return myLast + ", " + myFirst + " " + myMiddle.charAt(0); // charAt gets the character at an indexed location
   }

   // creates a new scanner for user input
   Scanner input = new Scanner(System.in);

   // this function reads user input and puts it into the name variables
   void read()
   {
      myFirst = input.nextLine(); // next line gets the next line of the user input
      myMiddle = input.nextLine();
      myLast = input.nextLine();
   }

  };

File: projects/08/NameTester.rb
# NameTester.rb tests class Name and its operations
#
# Begun by: Dr. Adams, for CS 214 at Calvin College.
# Completed by: Justin Voss
# Date: 3/10/22
####################################################

require 'test/unit/assertions'   # needed for assert
include Test::Unit::Assertions



class Name 
   # This function initializes the first, middle, and last names
   def initialize(first, middle, last)
      @first, @middle, @last = first, middle, last
    end

    # accessor method. By changing this to attr_accessor, it is now used for reading and writing
    attr_accessor :first, :middle, :last



    # This function concatenates the 3 parts of the name together
    def fullName
      @first + " " + @middle + " " + @last
    end

    # This function prints the name
    def print
      puts fullName # prints out the name
      fullName
    end

    # lfmi puts the name in last, first, middle initial format
    def lfmi
      return @last + ", " + @first + " " + @middle[0] + "."
    end

    # This function reads first, middle, last name from user input 
    def readName
      # gets.chomp gets a user inputted word
      @first = gets.chomp
      @middle = gets.chomp
      @last = gets.chomp
    end
end


def testName
   name = Name.new("John", "Paul", "Jones")

   assert name.first == "John", "first failed"
   assert name.middle == "Paul", "middle failed"
   assert name.last == "Jones", "last failed"
   assert name.fullName == "John Paul Jones", "fullName failed"
   assert name.print == "John Paul Jones", "print failed"

  # Testing Mutators
  name.first = "Chris"
  assert name.first == "Chris", "first failed"
  name.middle = "Robert"
  assert name.middle == "Robert", "first failed"
  name.last = "Evans"
  assert name.last == "Evans", "first failed"
  name.print
# Testing lfmi()
  puts name.lfmi
  assert name.lfmi == "Evans, Chris R.", "lfmi failed"
  print "All tests passed!\n"

  puts "Enter a first, middle, and last name: \n"
  name.readName
  name.print
end

testName

